[Constraint System] Implement heuristics for linked operator expressions in the solver proper.

hborla opened the pull request:
This PR contains the first handful of commits from @gregomni in https://github.com/apple/swift/pull/26140, with some adjustments. Specifically:
* If an operator overload choice has the same type as an existing overload binding elsewhere in the expression, attempt that choice first. This is done by ordering such choices first in the main disjunction partition rather than using the favoring mechanism, because the solver will stop if it successfully finds a solution with a favored constraint, which could be incorrect.
* Consider the number of favored choices when choosing which disjunction to try first in `selectDisjunction`.

These heuristics together help the solver reach a solution fast for expressions with multiple operators chained together where the operands all have the same type. 

Finding a solution fast allows the solver to do more aggressive pruning later on. Once this change lands, I'm going to work on pruning heuristics for generic overloads, which I think will finally let us remove `LinkedExprAnalyzer` and the hacks around favored constraints. Once those hacks are removed, the favoring mechanism can be used to suggest choices that are likely to work without worrying about correctness.

All thanks to Greg for the approach taken in this PR!!

gregomni committed:
Add condition for optimizing series of binops that every possible overload is (T,T)->T, with no (T,T)->U

gregomni committed:
Instead of chaining binops, favor disjunctions with op overloads whose types match existing binding choices

gregomni committed:
Favor operators based on existing binds via both basename and type

gregomni committed:
Merge typevars of operators with shared decls after finding a solution to speed up further searching

gregomni committed:
Dump disjunction possibilities on multiple aligned lines for better readability.

hborla committed:
[ConstraintSystem] Move getResolvedOverloads() from CSStep to ConstraintSystem.

hborla committed:
[ConstraintSystem] Fix build failures and formatting.

hborla committed:
[ConstraintSystem] Fix build failures.

hborla committed:
[ConstraintSystem] Allow the solver to find other solutions after
successfully finding a solution by favoring operators already bound
elsewhere.

Favoring existing operator bindings often lets the solver find a solution
fast, but it's not necessarily the best solution.

hborla committed:
[Test] Adjust type checker performance tests

hborla commented:
@swift-ci please smoke test compiler performance

hborla commented:
@swift-ci please test compiler performance

hborla committed:
[ConstraintSystem] Instead of favoring existing operator bindings,
order them first in the main disjunction partition.

hborla committed:
[ConstraintSystem] Remove an unnecessary SIMD special case in
`DisjunctionStep::shortCircuitDisjunctionAt`.

This code is unnecessary because SIMD overloads are in their own
partition, so the short circuiting will happen automatically.

hborla committed:
[ConstraintSystem] Re-instate the operator type variable merging
hacks for now.

There's some more disjunction pruning work to be done before we can
remove this.

hborla commented:
@swift-ci please smoke test compiler performance

hborla commented:
@swift-ci please test compiler performance

hborla committed:
[NFC][ConstraintSystem] Update the documentation for
existingOperatorBindingsForDisjunction.

hborla commented:
@swift-ci please test compiler performance

hborla commented:
@swift-ci please test source compatibility

hborla commented:
@swift-ci please test

hborla commented:
@swift-ci please test

hborla commented:
@swift-ci please test compiler performance

xedin reviewed (commented):
Makes sense to me! I have left a couple of comments inline and I wonder whether performance regressions are related to the fact that solver would pull generic overloads up before attempting the ones with concrete types which means that it ends up exploring a lot more space than before. It'd be great to get them added to the suite.

hborla commented on lib/Sema/CSSolver.cpp:
```diff
@@ -2013,6 +2013,51 @@ static Constraint *tryOptimizeGenericDisjunction(
+      continue;
+
+    auto *decl = constraint->getOverloadChoice().getDecl();
+    if (typesFound.count(decl->getInterfaceType()->getCanonicalType()))
```
Ah, that's a good point that I hadn't thought of. I do think it only makes sense to do this if the requirements are the same...or maybe we should just restrict this heuristic to the exact same decl in the case of generics. I think the "same type" piece is most useful for concrete overloads

hamishknight replied:
Wouldn't the interface type be written in terms of `GenericFunctionType` in that case, which should include the requirements?

That being said, an identity check on the decl may be cheaper than computing the canonical type for a `GenericFunctionType` – as IIRC that may involve spinning up a new GSB (which was the motivation behind https://github.com/apple/swift/pull/32034).

hborla replied:
> Wouldn't the interface type be written in terms of GenericFunctionType in that case, which should include the requirements?

I know `GenericFunctionType` stores the generic signature which has the requirements, but I'm not sure whether comparing canonical types factors in requirements.

> That being said, an identity check on the decl may be cheaper than computing the canonical type for a GenericFunctionType 

Yeah, I think I'm just going to stick to a `ValueDecl` check in the case of generics. I think it's a lot more common to have mixed operator expressions that use concrete overloads of the same type, so this heuristic for generics is probably more work than it's worth.

xwu commented:
@hborla, I just want to put on your radar the following issue if you weren't already aware of it (I forget which SR was filed about it, but I think there is one). It would have been fixed by "designated operators," which is why I'm rather sad it didn't make it:

SE-0104 added heterogeneous comparison and bit shift operators for integer types in order to improve the user experience. These heterogeneous operators behave as intended with concrete types. However, integer comparisons in generic algorithms behave differently (and were source breaking too):

If one operand in a generic context is a literal, overload resolution will favor heterogeneous comparison with `IntegerLiteralType` over homogeneous comparison.

The same problem would occur in a nongeneric context but for two hacks: (1) to improve type-checking performance, the compiler won’t traverse the protocol hierarchy to rank all overloads of an operator function if it finds a matching overload defined in the concrete type; (2) to favor homogeneous comparison over heterogeneous comparison, [otherwise redundant concrete implementations](https://github.com/apple/swift/pull/9909) have been added to built-in numeric types.

This issue was encountered during review of the standard library’s implementation of `DoubleWidth`, which in fact [had a bug](https://github.com/apple/swift/pull/9367#discussion_r118612475) as a consequence of the following behavior:

```swift
func f() -> Bool {
  return UInt.max == ~0
}
func g<T : FixedWidthInteger>(_: T.Type) -> Bool {
  return T.max == ~0
}

f()          // `true`
g(UInt.self) // `false`
```

In a generic context, even if the most refined protocol implements its own overload of the homogeneous comparison operator, Swift will look for all overloads of the operator by traversing the entire protocol hierarchy. Since heterogeneous comparison operators are defined somewhere along the hierarchy, the compiler will always find an overload that takes the “preferred” `IntegerLiteralType` and therefore infers the literal to be of type `Int`. As a result, in the expression `g(UInt.self)`, we are actually comparing `UInt.max` to `~(0 as Int)`.

Heterogeneous comparison were also planned for floating-point types. However, until the issue described above is resolved, adding heterogeneous comparison would cause new unexpected results:

```swift
func isOnePointTwo<T: BinaryFloatingPoint>(_ value: T) -> Bool {
  return value == 1.2
}

isOnePointTwo(1.2 as Float80)
// Currently `true`.
// Would be `false` when heterogeneous comparison is implemented.
```

Even before the constraints of ABI stability, I had tried a number of solutions at the level of the standard library to get the compiler to prefer homogeneous comparison, but it became apparent that there was no solution without tackling operator expressions in the constraint system.

hborla committed:
[ConstraintSystem] Only factor in existing operator bindings in
selectDisjunction if both disjunctions are operator bindings.

hborla committed:
[ConstraintSystem] For generic operators, only consider exact decls that
are already bound elsewhere for existingOperatorBindingsForDisjunction,
rather than also considering overload choices with the same type.

hborla committed:
[NFC][ConstraintSystem] Use llvm::count_if for the count methods on
Constraint.

hborla commented:
@swift-ci please smoke test compiler performance

hborla commented:
@swift-ci please test compiler performance

hborla commented:
@swift-ci please build toolchain

hborla commented:
@swift-ci please test source compatibility

hborla commented:
@swift-ci please smoke test compiler performance

hborla commented:
@swift-ci please test compiler performance

hamishknight commented on lib/Sema/CSSolver.cpp:
```diff
@@ -2013,6 +2013,60 @@ static Constraint *tryOptimizeGenericDisjunction(
+    return;
+
+  SmallSet<CanType, 4> concreteTypesFound;
+  SmallSet<ValueDecl *, 4> genericDeclsFound;
```
Would it be worth adding a comment to explain the difference in handling for generic overloads?

hborla replied:
Yes, you're right that this needs an explanation. I added a comment - let me know if you think it needs more clarification!

hamishknight replied:
Looks good to me!

hamishknight reviewed (commented):
So happy to see more of these solving hacks removed!

hborla committed:
[Test] Add a slow type checker test case from source compatibility suite.

hborla commented on validation-test/Sema/type_checker_perf/slow/mixed_string_array_addition.swift:
```diff
@@ -0,0 +1,12 @@
+func method(_ arg: String, body: () -> [String]) {}
+
+func test(str: String, properties: [String]) {
+  // expected-error@+1 {{the compiler is unable to type-check this expression in reasonable time}}
```
This isn't new, but a case I discovered when debugging the source compat failures. This compiles (very slowly) if you add a type annotation to `param` below.

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please test compiler performance

hborla commented:
@swift-ci please smoke test OS X platform

hborla commented:
@swift-ci please smoke test
